org/mozilla/jss/crypto/JSSOAEPParameterSpec.java

```diff
@@ -0,0 +1,196 @@
+            throw new IllegalArgumentException(msg);
+        }
+
+        // Do nothing. We just validate this data so if we get passed
+        // something unexpected, we error out instead.
+    }
+
+    public void setMaskGenAlgorithmType(String algo) throws IllegalArgumentException {
+        switch (algo.toLowerCase()) {
+            case "sha1":
+            case "sha-1":
+            case "ckm_sha_1":
+                hashAlg = PKCS11Constants.CKG_MGF1_SHA1;
+                break;
+            case "sha256":
+            case "sha-256":
+            case "ckm_sha256":
+                hashAlg = PKCS11Constants.CKG_MGF1_SHA256;
+                break;
+            case "sha384":
+            case "sha-384":
+            case "ckm_sha384":
+                hashAlg = PKCS11Constants.CKG_MGF1_SHA384;
+                break;
+            case "sha512":
+            case "sha-512":
+            case "ckm_sha512":
+                hashAlg = PKCS11Constants.CKG_MGF1_SHA512;
+                break;
+            default:
+                String msg = "Unknown mask generation algorithm identifier: " + algo;
+                throw new IllegalArgumentException(msg);
+        }
+    }
+
+    public void setMaskGenAlgorithmType(AlgorithmParameterSpec algo) throws IllegalArgumentException {
+        if (!(algo instanceof MGF1ParameterSpec) || algo == null) {
+            String msg = "Unknown mask generation algorithm parameter ";
+            msg += "specification: " + algo;
+            throw new IllegalArgumentException(msg);
+        }
+
+        MGF1ParameterSpec mgf1 = (MGF1ParameterSpec) algo;
+        switch (mgf1.getDigestAlgorithm().toLowerCase()) {
+            case "sha1":
+            case "sha-1":
+                mgf = PKCS11Constants.CKG_MGF1_SHA1;
+                break;
+            case "sha256":
+            case "sha-256":
+                mgf = PKCS11Constants.CKG_MGF1_SHA256;
+                break;
+            case "sha384":
+            case "sha-384":
+                mgf = PKCS11Constants.CKG_MGF1_SHA384;
+                break;
+            case "sha512":
+            case "sha-512":
+                mgf = PKCS11Constants.CKG_MGF1_SHA512;
+                break;
+            default:
+                String msg = "Unknown mask generation algorithm identifier: ";
+                msg += mgf1.getDigestAlgorithm();
+                throw new IllegalArgumentException(msg);
+        }
+    }
+
+    public void setMaskGenAlgorithmType(long algo) throws IllegalArgumentException {
+        mgf = algo;
+    }
+
+    public void setPSource(PSource spec) throws IllegalArgumentException {
+        if (spec == null) {
+            return;
+        }
+
+        if (!(spec instanceof PSource.PSpecified)) {
+            String msg = "Expected PSource spec to be an instance of ";
+            msg += "PSource.PSpecified, but wasn't: " + spec;
+            throw new IllegalArgumentException(msg);
+        }
```

Comment:
Is this necessary? Java will throw `ClassCastException` automatically in the following code.

Reply:
I think the `IllegalArgumentException` is nicer than the `ClassCastException` personally in this case.